### ISSUE TITLE
Details polyfill doesn't open/close with spacebar/enter key when JAWS is running

### ISSUE THREAD

**Issue by gavboulton** [alphagov/govuk_elements#130]
In user testing we've seen a partially sighted user unable to open the details element when using JAWS screen reader. The summary is described as a button, when pressing the spacebar/enter key, the key name is read out but the element never opens.

JAWS seems to prevent the spacebar/enter key from being captured, unless another key is pressed before. So, if you press the Windows key and then spacebar the details element does open.

I've been able to replicate this problem using Vista/IE9/JAWS 13. I've tried adding `aria-pressed` attributes to the summary element and have tried replacing `role="group"` with `role="application"` but neither of these have worked.

**Comment by edwardhorsford** [comment]
@aduggin you might want to take a look at this?

**Comment by robinwhittleton** [comment]
This seems to be specific to JAWS on IE.

**Comment by selfthinker** [comment]
This also came up during our testing of Elements. @accessiblewebuk found that although JAWS and NVDA work fine in other browsers, both cannot access the details in IE 11 (and probably older IEs). It is keyboard-accessible when JAWS and NVDA are switched off, but not when one of them is switched on.

I started looking at this today and found the reason. Some screen readers map the screen reader key + space (or enter) to the **click** event. As you can see in [Event Handlers and Screen Readers](http://unobfuscated.blogspot.co.uk/2013/05/event-handlers-and-screen-readers.html), the click event is usually the **only** event fired for IE and JAWS/NVDA combinations.

Our polyfill is currently listening for the keyup, keypress and mouseup events. It used to listen to the click instead of the mouseup event, but that was [changed because it made it trigger twice in Chrome](https://github.com/alphagov/govuk_elements/commit/66ae6ce45a75a74fd4dcc61ecac8ad432a93b6b1). That's also the reason why I didn't simply create a PR as it doesn't seem to be that easy to know which events are the right ones.

I've had a quick look at some other accessible details polyfills:
* [better-details-polyfill](https://github.com/chemerisuk/better-details-polyfill) uses click and keydown
* [jquery-details](https://github.com/mathiasbynens/jquery-details) uses click and keyup

Now I hope to wiretap the hive mind here. (I'm not too experienced with JS events.) There must be some universal fix for this, right? Which events should trigger the opening of details?

If you'd like to play around with various events, I have [isolated all the important bits](http://jsbin.com/vucicis).

**Comment by selfthinker** [comment]
FYI, this is also an interesting relevant resource: [ARIA buttons with different event handlers](https://dequeuniversity.com/testsuite/buttons-links/event-handlers)

**Comment by 36degrees** [comment]
So if I'm understanding correctly, the issue was that pressing space in Chrome would also trigger the `click` event, so it'd fire twice because of that?

I wonder if calling `preventDefault` on the `keyup` event would be enough to stop Chrome from triggering a click event as well? In which case we could go back to `click` / `keydown`. I'd suggest trying `keydown` first because I _think_ that would also prevent the page from being scrolled?

**Comment by selfthinker** [comment]
I wonder if there was only a problem because we had `keyup` as well as `keydown` and not because of the `click`? I'll just try changing the `mouseup` to `click` and test in all the scenarios (starting with Chrome).

**Comment by selfthinker** [comment]
Hang on... Whatever we'll do, some events will always be triggered together. I think I know what to do...
